### PR TITLE
COOP: Test history navigation

### PR DESCRIPTION
**Pull request by zcorpan**
The top-level page has no COOP, and opens a popup that also has no COOP.
The popup navigates to a cross-origin page with COOP: same-origin.
That page then navigates back in history.

Assert that the final page doesn't have .opener and doesn't have DOM
access to the opener.

Both navigations create a new browsing context group per spec.

### CODE REVIEW

File: html/cross-origin-opener-policy/resources/coop-coep.py [hunk 1]
```diff
@@ -21,13 +21,41 @@ def main(request, response):
...
 <iframe></iframe>
 <script>
-  const navigate = new URL(location).searchParams.get("navigate");
-  if (navigate !== null) {
-    self.location = navigate;
+  const params = new URL(location).searchParams;
+  const navHistory = params.get("navHistory");
+  const avoidBackAndForth = params.get("avoidBackAndForth");
+  const navigate = params.get("navigate");
+  const openerDOMAccess = params.get("openerDOMAccess");
```
Comment (jugglinmike) [inline html/cross-origin-opener-policy/resources/coop-coep.py]: This constant is unused.

File: html/cross-origin-opener-policy/resources/coop-coep.py [hunk 2]
```diff
@@ -21,13 +21,41 @@ def main(request, response):
...
+  // Need to wait until the page is fully loaded before navigating
+  // so that it creates a history entry properly.
+  async function fullyLoaded() {
```
Comment (jugglinmike) [inline html/cross-origin-opener-policy/resources/coop-coep.py]: This can be a synchronous function. Or maybe better, just get rid of the function and store the Promise value directly.

File: html/cross-origin-opener-policy/resources/coop-coep.py [hunk 3]
```diff
@@ -21,13 +21,41 @@ def main(request, response):
...
+    return new Promise((resolve, reject) => {
```
Comment (jugglinmike) [inline html/cross-origin-opener-policy/resources/coop-coep.py]: Maybe `addEventListener('error', reject);`

File: html/cross-origin-opener-policy/resources/coop-coep.py [hunk 4]
```diff
@@ -21,13 +21,41 @@ def main(request, response):
...
+          });
+        });
+      });
+    })
+  }
+  if (navHistory !== null) {
+    fullyLoaded().then(() => {
+      history.go(Number(navHistory));
+    });
+  } else if (navigate !== null && (history.length === 1 || !avoidBackAndForth)) {
```
Comment (jugglinmike) [inline html/cross-origin-opener-policy/resources/coop-coep.py]: `avoidBackAndForth` doesn't seem necessary; I think the `history.length === 1` condition is sufficient.

**Review by jugglinmike — COMMENTED** [review]
Looking good, @zcorpan. I left some suggestions to improve the clarity. Some aspects of your design suggest support for more test cases--forward traversal comes to mind. Were you thinking of adding more in a follow-up patch?

File: html/cross-origin-opener-policy/resources/coop-coep.py (hunk 2 above)
Comment (zcorpan) [reply html/cross-origin-opener-policy/resources/coop-coep.py]: First I wanted to use `await fullyLoaded()` but then recalled that top-level `await` can't be used, so resorted to using `.then`. But feel free to tweak this 🙂

File: html/cross-origin-opener-policy/resources/coop-coep.py (hunk 3 above)
Comment (zcorpan) [reply html/cross-origin-opener-policy/resources/coop-coep.py]: `error` is happens for uncaught exceptions and syntax errors from parsing scripts. I don't think it's useful to reject the promise if that happens. `load` will happen anyway.

File: html/cross-origin-opener-policy/resources/coop-coep.py (hunk 4 above)
Comment (zcorpan) [reply html/cross-origin-opener-policy/resources/coop-coep.py]: It is, but it prevents other (future) tests from doing multiple chained navigations.

**Comment by zcorpan** [comment]
> Some aspects of your design suggest support for more test cases--forward traversal comes to mind.

I'm happy you noticed. 🙂

> Were you thinking of adding more in a follow-up patch?

Yeah, or at least make it easier to add more tests.

File: html/cross-origin-opener-policy/coop-navigated-history-popup.https.html [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<!doctype html>
+<title>Cross-Origin-Opener-Policy: a navigating popup that then goes back in history</title>
+<meta content=timeout value=long>
```
Comment (zcorpan) [inline html/cross-origin-opener-policy/coop-navigated-history-popup.https.html]: This is bogus

File: html/cross-origin-opener-policy/resources/coop-coep.py (hunk 1 above)
Comment (zcorpan) [reply html/cross-origin-opener-policy/resources/coop-coep.py]: I'm not sure why I thought I needed this. Removed. Though I'd like to check now how this impacts other tests...

Comment (zcorpan) [reply html/cross-origin-opener-policy/resources/coop-coep.py]: For now, I've skipped the assertion when `openerDOMAccess` is undefined.

**Comment by zcorpan** [comment]
I've addressed the comments, @jugglinmike

**Comment by zcorpan** [comment]
@mikewest could you (or someone in your team) review this?

File: html/cross-origin-opener-policy/resources/common.js [hunk 6]
```diff
@@ -2,12 +2,16 @@ const SAME_ORIGIN = {origin: get_host_info().HTTPS_ORIGIN, name: "SAME_ORIGIN"};
 const SAME_SITE = {origin: get_host_info().HTTPS_REMOTE_ORIGIN, name: "SAME_SITE"};
 const CROSS_ORIGIN = {origin: get_host_info().HTTPS_NOTSAMESITE_ORIGIN, name: "CROSS_ORIGIN"}
 
-function url_test(t, url, channelName, hasOpener) {
+function url_test(t, url, channelName, hasOpener, openerDOMAccess) {
```
Comment (ParisMeuleman) [inline html/cross-origin-opener-policy/resources/common.js]: Maybe this should be moved within a dedicated PR and update all the tests? Or is there a reason to add this specifically here?
Do we expect this to be different to hasOpener? When?

**Review by ParisMeuleman — APPROVED** [review]
LGTM overall, thanks!

File: html/cross-origin-opener-policy/resources/common.js (hunk 6 above)
Comment (zcorpan) [reply html/cross-origin-opener-policy/resources/common.js]: I think we should add this check to all tests, but as a follow-up.

It is different from `hasOpener` because opener can exist when there's no COOP and the opener and openee are not same-origin. Then DOM access is not allowed due to the same-origin policy.

**Comment by zcorpan** [comment]
["wpt-firefox-nightly-stability"](https://community-tc.services.mozilla.com/tasks/feZ9Kmj1R7mirugIjCdxcw/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FfeZ9Kmj1R7mirugIjCdxcw%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log) check failed for `coep-navigate-popup.https.html` but that test was not changed here. It's possible that a rebase fixes this, though.

["infrastructure/ tests"](https://community-tc.services.mozilla.com/tasks/feZ9Kmj1R7mirugIjCdxcw/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FfeZ9Kmj1R7mirugIjCdxcw%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log)
```
Unexpected Results
------------------
/infrastructure/assumptions/allowed-to-play.html
  CRASH /infrastructure/assumptions/allowed-to-play.html
```
I don't know what caused this. Seems unrelated, though.

**Comment by zcorpan** [comment]
I've rebased on latest master. Let's see what the results are for the stability check.

**Review by hemeryar — COMMENTED** [review]
Looks fine but asking myself the same question as @ParisMeuleman.
Edit: Just seen the comment, thanks lgtm.

**Comment by zcorpan** [comment]
This time `/html/cross-origin-opener-policy/historical/coep-navigate-popup-unsafe-inherit.https.html` was producing unstable results. But that test also isn't introduced or changed here. I can report an issue about these tests being flaky.

**Comment by zcorpan** [comment]
Filed https://github.com/web-platform-tests/wpt/issues/21107

I'll admin-merge this; it's reviewed and the new test appears stable.
